File: sidecar/reminders_sidecar/notifications.py
"""
Decides what to toast. The Rust side only fires what this returns.

Keeping the policy here means the sleep/wake behaviour is unit-testable without
a Windows box, a real clock, or an actual notification service.

The machine sleeping through due times is handled deliberately rather than
discovered in production:

  * Anything overdue by more than `stale_after_minutes` (default 60) is treated
    as missed while the machine was away. Those never produce individual
    toasts -- they collapse into one summary. Waking up to fourteen separate
    toasts for things due overnight is worse than useless.

  * Freshly-due reminders toast individually, but only up to
    `max_individual` (default 3). Beyond that they also collapse into a
    summary, so a batch of simultaneous due times can't flood the tray.

Every reminder handed back is marked notified in the same call, so a crash
between deciding and firing costs at most one toast rather than causing a loop.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timedelta, timezone
from typing import Optional

from .db import Cache, from_iso


@dataclass
class Toast:
    title: str
    body: str
    reminder_id: Optional[str] = None
    kind: str = "reminder"  # reminder | summary


@dataclass
class NotificationPlan:
    toasts: list[Toast] = field(default_factory=list)
    notified_ids: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        return {
            "toasts": [
                {
                    "title": t.title,
                    "body": t.body,
                    "reminder_id": t.reminder_id,
                    "kind": t.kind,
                }
                for t in self.toasts
            ],
            "notified_ids": self.notified_ids,
        }


def _fmt_due(iso: Optional[str]) -> str:
    dt = from_iso(iso)
    if dt is None:
        return ""
    return dt.astimezone().strftime("%a %d %b, %H:%M")


def plan_notifications(
    cache: Cache,
    now: Optional[datetime] = None,
    stale_after_minutes: int = 60,
    max_individual: int = 3,
) -> NotificationPlan:
    """Work out which toasts to fire this tick, and mark them notified."""
    now = now or datetime.now(timezone.utc)
    due = cache.due_unnotified(now)
    plan = NotificationPlan()
    if not due:
        return plan

    stale_cutoff = now - timedelta(minutes=stale_after_minutes)
    missed: list[dict] = []
    fresh: list[dict] = []
    for r in due:
        dt = from_iso(r.get("due_date"))
        if dt is not None and dt < stale_cutoff:
            missed.append(r)
        else:
            fresh.append(r)

    if missed:
        titles = ", ".join(r["title"] for r in missed[:3] if r.get("title"))
        more = len(missed) - 3
        if more > 0:
            titles += f", and {more} more"
        plan.toasts.append(
            Toast(
                title=f"{len(missed)} reminder{'s' if len(missed) != 1 else ''} "
                "were due while you were away",
                body=titles or "Open Reminders to review them.",
                kind="summary",
            )
        )
        plan.notified_ids += [r["id"] for r in missed]

    if fresh:
        if len(fresh) <= max_individual:
            for r in fresh:
                plan.toasts.append(
                    Toast(
                        title=r.get("title") or "Reminder",
                        body=_fmt_due(r.get("due_date")) or "Due now",
                        reminder_id=r["id"],
                    )
                )
        else:
            titles = ", ".join(r["title"] for r in fresh[:3] if r.get("title"))
            plan.toasts.append(
                Toast(
                    title=f"{len(fresh)} reminders are due",
                    body=titles,
                    kind="summary",
                )
            )
        plan.notified_ids += [r["id"] for r in fresh]

    if plan.notified_ids:
        cache.mark_notified(plan.notified_ids)
    return plan

File: sidecar/src/notifications.rs
use chrono::{DateTime, Duration, Utc};
use serde::Serialize;

use crate::cache::Cache;
use crate::error::Result;
use crate::timeutil::{local_zone, notify_at, parse_instant, to_iso};

#[derive(Debug, Serialize)]
pub struct Toast {
    pub title: String,
    pub body: String,
    pub reminder_id: Option<String>,
    pub kind: &'static str,
}

#[derive(Debug, Default, Serialize)]
pub struct NotificationPlan {
    pub toasts: Vec<Toast>,
    pub notified_ids: Vec<String>,
}

pub fn plan(
    cache: &Cache,
    now: DateTime<Utc>,
    stale_after_minutes: i64,
    max_individual: usize,
) -> Result<NotificationPlan> {
    let due = cache.due_unnotified(&to_iso(now))?;
    let cutoff = now - Duration::minutes(stale_after_minutes.clamp(1, 24 * 60));
    let mut missed = Vec::new();
    let mut fresh = Vec::new();
    for reminder in due {
        let Some(raw_due) = reminder.due_date.as_deref() else {
            continue;
        };
        let alert = notify_at(parse_instant(raw_due)?, reminder.all_day);
        if alert > now {
            continue;
        }
        if alert < cutoff {
            missed.push(reminder);
        } else {
            fresh.push(reminder);
        }
    }

    let mut result = NotificationPlan::default();
    if !missed.is_empty() {
        let mut names: Vec<_> = missed
            .iter()
            .take(3)
            .filter(|r| !r.title.is_empty())
            .map(|r| r.title.clone())
            .collect();
        if missed.len() > 3 {
            names.push(format!("and {} more", missed.len() - 3));
        }
        result.toasts.push(Toast {
            title: format!(
                "{} reminder{} were due while you were away",
                missed.len(),
                if missed.len() == 1 { "" } else { "s" }
            ),
            body: if names.is_empty() {
                "Open Reminders to review them.".into()
            } else {
                names.join(", ")
            },
            reminder_id: None,
            kind: "summary",
        });
        result
            .notified_ids
            .extend(missed.iter().map(|r| r.id.clone()));
    }

    if !fresh.is_empty() {
        if fresh.len() <= max_individual.clamp(1, 10) {
            for reminder in &fresh {
                let due = reminder
                    .due_date
                    .as_deref()
                    .map(parse_instant)
                    .transpose()?;
                let body = due
                    .map(|value| {
                        let local = value.with_timezone(&local_zone());
                        if reminder.all_day {
                            local.format("%a %d %b").to_string()
                        } else {
                            local.format("%a %d %b, %H:%M").to_string()
                        }
                    })
                    .unwrap_or_else(|| "Due now".into());
                result.toasts.push(Toast {
                    title: if reminder.title.is_empty() {
                        "Reminder".into()
                    } else {
                        reminder.title.clone()
                    },
                    body,
                    reminder_id: Some(reminder.id.clone()),
                    kind: "reminder",
                });
            }
        } else {
            result.toasts.push(Toast {
                title: format!("{} reminders are due", fresh.len()),
                body: fresh
                    .iter()
                    .take(3)
                    .map(|r| r.title.as_str())
                    .filter(|s| !s.is_empty())
                    .collect::<Vec<_>>()
                    .join(", "),
                reminder_id: None,
                kind: "summary",
            });
        }
        result
            .notified_ids
            .extend(fresh.iter().map(|r| r.id.clone()));
    }

    cache.mark_notified(&result.notified_ids)?;
    Ok(result)
}
